lib/mayaUsd/fileio/jobs/writeJob.cpp

```diff
@@ -261,49 +306,189 @@ class AutoUpAxisAndUnitsChanger : public MayaUsd::AutoUndoCommands
+    class JobFramesWriter
+    {
+    public:
+        JobFramesWriter() = delete;
+        explicit JobFramesWriter(UsdMaya_WriteJob* job);
+
+        bool Finished() const { return _next == _end; }
+        auto AllSamples() const -> const TimeSamples&;
+        bool WriteFrameIfNeeded(double frame);
+
+    private:
+        UsdMaya_WriteJob*           _job;
+        TimeSamples::const_iterator _next;
+        TimeSamples::const_iterator _end;
+    };
+    using JobFramesWriters = std::vector<JobFramesWriter>;
+
+    /// Computes the ordered union of all \p writers time-samples.
+    static TimeSamples _UnionTimeSamples(const JobFramesWriters& writers);
+};
+
+UsdMaya_WriteJobImpl::JobFramesWriter::JobFramesWriter(UsdMaya_WriteJob* job)
+    : _job(job)
+    , _next(job->mJobCtx.GetArgs().timeSamples.cbegin())
+    , _end(job->mJobCtx.GetArgs().timeSamples.cend())
+{
+}
+
+bool UsdMaya_WriteJobImpl::JobFramesWriter::WriteFrameIfNeeded(double frame)
+{
+    if (Finished())
+        return false;
+
+    if (frame != *_next)
+        return true;
+
+    ++_next;
+    return _job->_WriteFrame(frame);
+}
+
+auto UsdMaya_WriteJobImpl::JobFramesWriter::AllSamples() const -> const TimeSamples&
+{
+    return _job->mJobCtx.GetArgs().timeSamples;
+}
+
+// static
+auto UsdMaya_WriteJobImpl::_UnionTimeSamples(const JobFramesWriters& writers) -> TimeSamples
+{
+    if (writers.empty())
+        return {};
+
+    TimeSamples unioned = writers.front().AllSamples();
+
+    TimeSamples storage;
+    for (auto itr = writers.cbegin() + 1; itr != writers.cend(); ++itr)
+        UsdMayaWriteUtil::UpdateTimeSamples(unioned, itr->AllSamples(), &storage);
+
+    return unioned;
+}
+
+// static
+bool UsdMaya_WriteJobImpl::WriteJobs(const std::vector<UsdMaya_WriteJob*>& jobs)
 {
-    const std::vector<double>& timeSamples = mJobCtx.mArgs.timeSamples;
+    if (jobs.empty())
+        return true;
+
+    // Get the wanted upAxis, unit and renderLayer node.
+    const auto* firstJob = jobs.front();
+    const auto& firstArgs = firstJob->mJobCtx.GetArgs();
+
+    const auto usdUpAxis = _WantedUSDUpAxis(firstArgs.upAxis);
+    const auto usdMetersPerUnit = _WantedUSDMetersPerUnit(firstArgs.unit);
+    const auto renderLayer = _WantedRenderLayerNode(firstArgs.renderLayerMode);
+
+    // Validate that multiple jobs can be exported all together.
+    if (jobs.size() > 1) {
+        auto argsAreCompatible = [&](const UsdMayaJobExportArgs& args) -> bool {
+            return (_WantedUSDUpAxis(args.upAxis) == usdUpAxis)
+                && (_WantedUSDMetersPerUnit(args.unit) == usdMetersPerUnit)
+                && (_WantedRenderLayerNode(args.renderLayerMode) == renderLayer);
+        };
+
+        std::set<std::string> seenFileNames;
+        for (auto& job : jobs) {
+            // Verify that we are not writing twice to the same filename.
+            if (!seenFileNames.insert(job->_fileName).second) {
```

Comment:
Maybe in the future we could support same file name in append mode?

I guess it is safer to have different files for now, due to possible collisions.